### DISCUSSION TITLE
Add constraints for optimization problems

### DISCUSSION THREAD

**Discussion by JingyuQian** [facebookresearch/theseus#487] [Q&A] [answered]
## 🚀 Feature

Don't know if this has been raised before or should be the purpose of this lib, but it would be nice to allow adding constraints for an optimization problem.

## Motivation

Some optimization problems take on the form:

$$min\text{ } F(x,y,z),\text{         } s.t. \text{       }x\leq\alpha,...$$

while $F(x,y,z)$ is easy to implement as:

$$F(x,y,z)=\sum_i[w_ig_i(x,y,z)]^2$$

the constraints don't have an explicit way to implement. One alternative is to write it as a normal loss and give a relatively large $w_i$, but I found it in practice this doesn't work very well.

## Pitch

Something like the following

```python
objective = th.Objective()
# Add losses
objective.add_loss(loss_function)
...
# Add constraints
objective.add_constraint(constraint_function)
```

## Additional context

The project `casadi` has something as 

```python
from casadi import Opti
optim = Opti()
optim.subject_to(optim.bounded(lower_bound, x, upper_bound))
```

**Comment by luisenp** [comment, marked as answer]
Hi @JingyuQian. Thanks for your interest in Theseus and the feature request! 

We are in fact starting to work on this feature, and there is an experimental PR up #457 (only for equality constraints). It's still pretty early in development, but we are hoping to have something running within the next two months.

**Reply by tbanker** [reply]
With it having been some time, are these constraint handling features still planning to be added? Features for equality and inequality constraints would both be beneficial to my current line of work.

**Reply by bamos** [reply]
Hi @tbanker, we don't have plans on merging that PR in with full generality, but the implementation of the augmented Lagrangian method in there is working and could be adaptable to your specific needs. Let us know if you have any issues with derivatives there. And if you end up using that and want to get something more general into the official code we are open to discussing more!

**Comment by JingyuQian** [comment]
Thanks @luisenp ! For my use case, inequality constraints are required, but this is good news to hear. I'll close this and keep watching the PR.

**Comment by mhmukadam** [comment]
@JingyuQian Thanks for your interest! I am curious what application you have in mind to leverage inequality constraints?

We have indeed received feature request for inequality constraints from many users (one of the most commonly requested ones), but given our team's bandwidth it will take us longer to prioritize and work on it. In the meantime, we of course welcome any community support :)

**Reply by JingyuQian** [reply]
@mhmukadam For the moment, I'm using Theseus purely as an optimization tool instead of a differentiable layer in a neural network, leveraging its ability to handle batched inputs. The problem involves planning a trajectory for a robot that observes all the kinematic rules, so there will be both equality and inequality constraints. As an example, the acceleration must not exceed a threshold. 

I can surely take a look into it and try to implement it when available, but I don't know where to start. Would you kindly point me in a direction?

**Reply by mhmukadam** [reply]
Good to know about use cases that want to leverage the batching capability!

That's great if you have a chance to look into it. Might be reasonable to start with the closest related ideas that involve inequality constraints on top of the current work in the Augmented Lagrangian PR. Maybe interior point method (for example see Calipso https://arxiv.org/pdf/2205.09255.pdf section 3.5) or Convex optimization (for instance see cvxpy layers https://arxiv.org/pdf/1910.12430.pdf).

Adding @bamos to this thread who is the expert in this space and can give better pointers and guidance.

**Comment by JingyuQian** [comment]
Another thing I'm curious about is, will the added constraints (both equality/inequality) affect the differentiability of a Theseus layer?

**Reply by mhmukadam** [reply]
Our aim will be to continue maintaining full differentiability and parallelization as we extend support to new optimizers.
